pkgs/_macro_host/lib/macro_host.dart

```diff
@@ -2,26 +2,91 @@
 // for details. All rights reserved. Use of this source code is governed by a
 // BSD-style license that can be found in the LICENSE file.
 
+import 'dart:io';
+
 import 'package:_macro_builder/macro_builder.dart';
 import 'package:_macro_runner/macro_runner.dart';
 import 'package:_macro_server/macro_server.dart';
+import 'package:dart_model/dart_model.dart';
 import 'package:macro_service/macro_service.dart';
 
-class MacroHost {
+/// Hosts macros: builds them, runs them, serves the macro service.
+///
+/// Tools that want to support macros, such as the Analyzer and the CFE, can
+/// do so by running a `MacroHost` and providing their own `MacroService`.
+class MacroHost implements MacroService {
   final MacroServer macroServer;
+  final ListOfServices services;
```

Comment:
This seems to be more of a higher level concept, sort of a hierarchical registry of services? I think there is probably a better name, maybe `ServiceRegistry serviceRegistry`? That would convey better that this is one service, with a pointer back to the registry of all services.

Or possibly restructure this so it doesn't add itself in the constructor, and instead in `serve` below you add it manually, I think I probably prefer that option (it doesn't seem like this list of services is needed after the constructor so keeping a back reference around after that wrong).

Another option could be just giving each `MacroService` a `MacroService? child;` field, and dropping the list of services wrapper entirely. The default implementation of `handle` would just call it on the child or throw. And each service would be required to call `super.handle(...)` if they don't handle a request.

Reply:
👍 Fine to follow up in later prs